### PR TITLE
Migrate /pkg/controller/disruption to structured and contextual logging

### PR DESCRIPTION
**Pull request by oxxenix**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
Switch the disruption controller code to use contextual and structured logging.

#### Which issue(s) this PR fixes:
Fixes #
Part of https://github.com/kubernetes/enhancements/issues/3077

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```release-note
Migrated the disruption controller (within `kube-controller-manager`) to use [contextual logging](https://k8s.io/docs/concepts/cluster-administration/system-logs/#contextual-logging).
```

/wg structured-logging
/area logging
/priority important-longterm
/kind cleanup
/cc @kubernetes/wg-structured-logging-reviews

### CODE REVIEW

File: pkg/controller/disruption/disruption.go [hunk 1]
```diff
@@ -469,45 +483,45 @@ func (dc *DisruptionController) removeDb(obj interface{}) {
 		}
 		pdb, ok = tombstone.Obj.(*policy.PodDisruptionBudget)
 		if !ok {
-			klog.Errorf("Tombstone contained object that is not a pdb %+v", obj)
+			klog.Errorf("Tombstone contained object that is not a PDB %+v", obj)
```
Comment (obaranov1) [inline pkg/controller/disruption/disruption.go:486]: @pohly, logger.Error requires err, so how we should proceed with klog.Errorf if no err here?

**Comment by obaranov1** [comment]
@pohly, not sure how to request review

File: pkg/controller/disruption/disruption.go (hunk 1 above)
Comment (pohly) [reply pkg/controller/disruption/disruption.go:486]: The `err` parameter may be nil: `logger.Error(nil, "Tombstone contained object that is not a PDB", "obj", obj)`.

**Comment by pohly** [comment]
> not sure how to request review

You can `/assign @pohly` (or anyone else).

File: pkg/controller/disruption/disruption.go [hunk 2]
```diff
@@ -446,20 +460,20 @@ func (dc *DisruptionController) Run(ctx context.Context) {
 	<-ctx.Done()
 }
 
-func (dc *DisruptionController) addDb(obj interface{}) {
+func (dc *DisruptionController) addDb(logger klog.Logger, obj interface{}) {
 	pdb := obj.(*policy.PodDisruptionBudget)
-	klog.V(4).Infof("add DB %q", pdb.Name)
-	dc.enqueuePdb(pdb)
+	logger.V(4).Info("Add DB", "PodDisruptionBudget", klog.KObj(pdb))
+	dc.enqueuePdb(logger, pdb)
```
Comment (pohly) [inline pkg/controller/disruption/disruption.go:466]: Watch out for unnecessary spaces.

Run `hack/update-gofmt.sh` when you editor doesn't handle this.

But ultimately you should better configure your editor to work well with Go, which typically includes live syntax checking and running "gofmt" on saving the file.

File: pkg/controller/disruption/disruption.go [hunk 3]
```diff
@@ -526,51 +540,51 @@ func (dc *DisruptionController) deletePod(obj interface{}) {
 			return
 		}
 	}
-	klog.V(4).Infof("deletePod called on pod %q", pod.Name)
-	pdb := dc.getPdbForPod(pod)
+	logger.V(4).Info("DeletePod called on pod", "Pod", klog.KObj(pod))
```
Comment (pohly) [inline pkg/controller/disruption/disruption.go:543]: "Pod" -> "pod"

Comment (pohly) [reply pkg/controller/disruption/disruption.go:543]: Also elsewhere...

File: pkg/controller/disruption/disruption.go [hunk 4]
```diff
@@ -516,60 +530,60 @@ func (dc *DisruptionController) deletePod(obj interface{}) {
...
 
-func (dc *DisruptionController) getPdbForPod(pod *v1.Pod) *policy.PodDisruptionBudget {
+func (dc *DisruptionController) getPdbForPod(logger klog.Logger, pod *v1.Pod) *policy.PodDisruptionBudget {
 	// GetPodPodDisruptionBudgets returns an error only if no
 	// PodDisruptionBudgets are found.  We don't return that as an error to the
 	// caller.
 	pdbs, err := dc.pdbLister.GetPodPodDisruptionBudgets(pod)
 	if err != nil {
-		klog.V(4).Infof("No PodDisruptionBudgets found for pod %v, PodDisruptionBudget controller will avoid syncing.", pod.Name)
+		logger.V(4).Info("No PodDisruptionBudgets found for pod, PodDisruptionBudget controller will avoid syncing.", "Pod", klog.KObj(pod))
```
Comment (pohly) [inline pkg/controller/disruption/disruption.go:586]:
```suggestion
		logger.V(4).Info("No PodDisruptionBudgets found for pod, PodDisruptionBudget controller will avoid syncing.", "pod", klog.KObj(pod))
```

File: pkg/controller/disruption/disruption.go [hunk 5]
```diff
@@ -705,11 +721,8 @@ func (dc *DisruptionController) trySync(ctx context.Context, pdb *policy.PodDisr
 	}
 	// We have unmamanged pods, instead of erroring and hotlooping in disruption controller, log and continue.
 	if len(unmanagedPods) > 0 {
-		klog.V(4).Infof("found unmanaged pods associated with this PDB: %v", unmanagedPods)
-		dc.recorder.Eventf(pdb, v1.EventTypeWarning, "UnmanagedPods", "Pods selected by this PodDisruptionBudget (selector: %v) were found "+
-			"to be unmanaged or managed by an unsupported controller (missing a /scale subresource). As a result, the status of the PDB "+
-			"cannot be calculated correctly, which may result in undefined behavior. To account for these pods please set \".spec.minAvailable\" "+
-			"field of the PDB to an integer value.", pdb.Spec.Selector)
+		logger.V(4).Info("Found unmanaged pods associated with this PDB:",
+			strings.Join(unmanagedPods, ",'"))
```
Comment (pohly) [inline pkg/controller/disruption/disruption.go:725]: Try running "hack/verify-golangci-lint.sh", it should complain about this (assuming that the base of your branch is recent enough).

```suggestion
		logger.V(4).Info("Found unmanaged pods associated with this PDB", "pods", unmanagedPods)
```

File: pkg/controller/disruption/disruption.go [hunk 6]
```diff
@@ -915,8 +929,8 @@ func (dc *DisruptionController) buildDisruptedPodMap(pods []*v1.Pod, pdb *policy
 		}
 		expectedDeletion := disruptionTime.Time.Add(DeletionTimeout)
 		if expectedDeletion.Before(currentTime) {
-			klog.V(1).Infof("Pod %s/%s was expected to be deleted at %s but it wasn't, updating pdb %s/%s",
-				pod.Namespace, pod.Name, disruptionTime.String(), pdb.Namespace, pdb.Name)
+			logger.V(1).Info("Pod was expected to be deleted but it wasn't, updating PDB",
+				"pod", klog.KObj(pod), "deletionTime", disruptionTime.String(), "PodDisruptionObject", klog.KObj(pdb))
```
Comment (pohly) [inline pkg/controller/disruption/disruption.go:933]:
```suggestion
				"pod", klog.KObj(pod), "deletionTime", disruptionTime, "PodDisruptionObject", klog.KObj(pdb))
```

**Review by pohly — CHANGES_REQUESTED** [review]

File: pkg/controller/disruption/disruption.go [hunk 7]
```diff
@@ -134,6 +134,7 @@ type controllerAndScale struct {
 type podControllerFinder func(ctx context.Context, controllerRef *metav1.OwnerReference, namespace string) (*controllerAndScale, error)
 
 func NewDisruptionController(
+	ctx context.Context,
```
Comment (pohly) [inline pkg/controller/disruption/disruption.go:137]: How can Kubernetes compile when the function requires a new parameter and some callers of this function aren't getting updated?

/test pull-kubernetes-verify-strict-lint

**Comment by pohly** [comment]
@yuzhiquan: can you also review the PR?

**Review by pohly — CHANGES_REQUESTED** [review]
Can you remove the exception from https://github.com/kubernetes/kubernetes/blob/5f23f835879b514e9bf16809059b38957d20a486/hack/logcheck.conf#L44?

**Comment by pohly** [comment]
The unit test failures seem genuine. Could it be that the content of the events that these tests seem to check for changed?

@obaranov1 can you investigate? Ping me on chat if you need help.

**Comment by pohly** [comment]
/close

Work continues in https://github.com/kubernetes/kubernetes/pull/119147

**Pull request closed without merging**